mmengine/hooks/checkpoint_hook.py

```diff
@@ -316,6 +320,10 @@ def _save_checkpoint(self, runner) -> None:
         Args:
             runner (Runner): The runner of the training process.
         """
+        if (self.by_epoch and runner.epoch < self.save_start) or (
+                not self.by_epoch and runner.iter < self.save_start):
+            return
```

Comment:
should this code be added to `after_train_iter` and `after_train_epoch`?

Reply:
I found that `after_train_iter` function is placed at the end of the `checkpoint_hook.py`, why not put it after the `after_val_epoch`, at least they have similar functions

Reply:
In my opinion, the `every_n_train_iters` and `every_n_epochs` functions also need to be rewritten, because with `save_start` you should think `save_start` with `self.interval`

Reply:
> I found that `after_train_iter` function is placed at the end of the `checkpoint_hook.py`, why not put it after the `after_val_epoch`, at least they have similar functions

Agree :rofl:, I think `after_train_iter` is not placed in the correct position either. 

I think we should keep the single responsibility of `every_n_iter` and `every_n_epoch`.  Their functions should be the same as their name expressed. If we want to add an extra check, like `save_start`, will the following way more proper?

```python
    def after_train_iter(self,
                         runner,
                         batch_idx: int,
                         data_batch: DATA_BATCH = None,
                         outputs=Optional[dict]) -> None:
        if self.by_epoch or runner.iter < self.save_start:
            return
        # save checkpoint for following cases:
        # 1. every ``self.interval`` iterations
        # 2. reach the last iteration of training
        if self.every_n_train_iters(runner, self.interval) or \
                (self.save_last and
                 self.is_last_train_iter(runner)):
            runner.logger.info(
                f'Saving checkpoint at {runner.iter + 1} iterations')
            self._save_checkpoint(runner)
```

Reply:
But if you set `save_start` and `interval` at the same time, then it should start from `save_start` and save the weights once every `interval`? So I would think that `every_n_train_iters` and `every_n_epochs` code should be modified to meet `save_start` scenarios

Reply:
Got it, it makes sense to me.